src/Processors/Formats/IOutputFormat.cpp

```diff
@@ -205,7 +218,8 @@ void IOutputFormat::onProgress(const Progress & progress)
         if (elapsed_ns >= prev_progress_write_ns + 1000 * progress_write_frequency_us)
         {
             std::unique_lock lock(writing_mutex, std::try_to_lock);
-            if (lock)
+
+            if (lock && has_progress_update_to_write && !finalized)
```

Comment:
Does it make sense to check `has_progress_update_to_write` that was set to true a few lines above?
If it can be concurrently changed before the lock is taken, why don't we update it with the lock held?
All other methods here, except explicit `finalizeUnlocked` update this flag with lock